### DISCUSSION TITLE
Things to be improved

### DISCUSSION THREAD

**Discussion by stevegalante** [apohl79/audiogridder#472] [Help]
Using Audiogridder on the same machine in aax format (ProTools Ultimate 2021.3.0 on Macpro running Mojave 10.14.6), 
and I noticed the following : 
- when you click on a plugin name instantiated in AG the GUI is always shown on top-left of the screen. If you move it but it's still within the ProTools window and you go back to ProTools it disappears, and when you click on the plugin name it reappears always on the top left. if you move it out of ProTools window then it stays there, but again if you accidentally click on the plugin name it moves to the top left again.
- when you open a plugin GUI and you tweak the plugin you lose all ProTools functions, even the basic spacebar for Play/Stop. This does not happen, for example, with a similar hosting software like Bluecat Patchwork.
- when you instantiate a plugin in ProTools in multi-mono, and the plugin has a long latency (like with some Acustica Audio plugins) there's an audible time difference between the left and right channel, or sometimes one of the two channels is muted. 
- sometimes ProTools crashes when removing some plugin, even though I haven't been able to exactly reproduce this issue.
Thank you for your attention and congrats on the development of this plugin !

**Comment by apohl79** [comment]
> Using Audiogridder on the same machine in aax format (ProTools Ultimate 2021.3.0 on Macpro running Mojave 10.14.6),
> and I noticed the following :
> 
> * when you click on a plugin name instantiated in AG the GUI is always shown on top-left of the screen. If you move it but it's still within the ProTools window and you go back to ProTools it disappears, and when you click on the plugin name it reappears always on the top left. if you move it out of ProTools window then it stays there, but again if you accidentally click on the plugin name it moves to the top left again.

1.2 will have a "local" mode. It will help a bit and will try to stick the plugin UI window to the AG plugin window. It will also allow for parallel plugin UIs.

> * when you open a plugin GUI and you tweak the plugin you lose all ProTools functions, even the basic spacebar for Play/Stop. This does not happen, for example, with a similar hosting software like Bluecat Patchwork.

This is not easy to solve at the moment. Patchwork - as far as I can see - is a plugin that can directly host other plugins (within the DAW context). AG is a network bridge. That means plugins are always hosted in a separate process outside the DAW (usually running somewhere, and in this special case locally), who owns the windows etc. If you work in the plugin UI, you are not working in the DAW at that point and keyboard input ends up in AG server, rather than the DAW. I tried to send a signal to the AG plugin, once you press space, but the plugin can't control the host transport usually (I tried logic and cubase, but I see no indication in the AAX code, that it would work in pro tools).

> * when you instantiate a plugin in ProTools in multi-mono, and the plugin has a long latency (like with some Acustica Audio plugins) there's an audible time difference between the left and right channel, or sometimes one of the two channels is muted.

This needs more investigation. Created issue https://github.com/apohl79/audiogridder/issues/477.

> * sometimes ProTools crashes when removing some plugin, even though I haven't been able to exactly reproduce this issue.
>   Thank you for your attention and congrats on the development of this plugin !

Once you can reproduce this, please create a bug report and please make sure tracing is enabled. https://audiogridder.com/bug-reports/

**Reply by stevegalante** [reply]
Great news, thank you !

**Reply by stevegalante** [reply]
> This is not easy to solve at the moment. Patchwork - as far as I can see - is a plugin that can directly host other plugins (within the DAW context). AG is a network bridge. That means plugins are always hosted in a separate process outside the DAW (usually running somewhere, and in this special case locally), who owns the windows etc. If you work in the plugin UI, you are not working in the DAW at that point and keyboard input ends up in AG server, rather than the DAW. I tried to send a signal to the AG plugin, once you press space, but the plugin can't control the host transport usually (I tried logic and cubase, but I see no indication in the AAX code, that it would work in pro tools).
> 
You can have a look at how Vienna Ensemble Pro works. It's also a network system - separate application - but you can use the spacebar from it and it works on ProTools (don't know how it does though)

**Reply by emack999** [reply]
> * when you open a plugin GUI and you tweak the plugin you lose all ProTools functions, even the basic spacebar for Play/Stop. This does not happen, for example, with a similar hosting software like Bluecat Patchwork.

This is why you're better off using host automation,  Select the automation parameters inside Autogridder, then map to your DAW and control those parameters from there without ever opening the plugin itself (and in your case, avoid losing your ProTools functions).

This is a good lesson for workflow in general, and specifically good when wanting to modify plugin parameters enveloped by Autogridder,  In reality, you don't want to be opening plugin windows all over the place for many reasons. This is what host automation is for.

**Reply by apohl79** [reply]
> You can have a look at how Vienna Ensemble Pro works. It's also a network system - separate application - but you can use the spacebar from it and it works on ProTools (don't know how it does though)

So you can press space _in_ vep when working in the plugin loaded into vep (like on a different computer) and it starts playback in pt?

**Reply by apohl79** [reply]
> This is why you're better off using host automation,  Select the automation parameters inside Autogridder, then map to your DAW and control those parameters from there without ever opening the plugin itself (and in your case, avoid losing your ProTools functions).

fyi, you can also use the generic editor in AG to avoid the plugin UI.

**Reply by apohl79** [reply]
> > You can have a look at how Vienna Ensemble Pro works. It's also a network system - separate application - but you can use the spacebar from it and it works on ProTools (don't know how it does though)
> 
> So you can press space _in_ vep when working in the plugin loaded into vep (like on a different computer) and it starts playback in pt?

Never mind, found a way on macOS at least to forward keystrokes into non focused apps. Likely working similarly on Windows.

**Reply by stevegalante** [reply]
That would be a huge improvement ! Yes from the VEP you can make the spacebar work in ProTools while focus is still on the VEP. 
Don't fully understand what emack999 means by mapping and avoiding the plugin GUI...

**Reply by apohl79** [reply]
yep, works on macOS and Windows now. will be in the next beta.

regarding automation, you can control plugin params without the plugin UI via the generic editor or host automation. that way you would not leave the DAW context.

**Reply by emack999** [reply]
> Don't fully understand what emack999 means by mapping and avoiding the plugin GUI...

@stevegalante You can select parameters of the plugin wrapped by AG by right-clicking on the plugin name, scrolling down to Automation, then choose the parameter(s) you want to expose to ProTools.
https://imgur.com/a/ldB0rFl

Then when you create your automation lanes in ProTools, the plugin parameter(s) you chose in the above steps should be listed.

Voila!  You can now adjust the parameters of that plugin without having to open the plugin itself or the AG wrapper.
